Clean up wolfSSL_clear() and add some more logging

julek-wolfssl opened the pull request:
Implement WOLFSSL_MSG_EX() which uses XVSNPRINTF to allow for formatted strings to be printed. It uses a 100 byte internal stack buffer to format the log message.

julek-wolfssl committed:
Free session in `wolfSSL_clear`

- Define `NO_SESSION_CACHE_REF` in wpa build

julek-wolfssl committed:
Add some missing logs and implement WOLFSSL_MSG_EX()

WOLFSSL_MSG_EX() uses XVSNPRINTF to allow for formatted strings to be printed. It uses a 100 byte internal stack buffer to format the log message.

dgarske commented on wolfcrypt/src/logging.c:
```diff
@@ -364,6 +364,27 @@ static void wolfssl_log(const int logLevel, const char *const logMessage)
+void WOLFSSL_MSG_EX(const char* fmt, ...)
+{
+    if (loggingEnabled) {
+        char msg[100];
```
Consider making `100` an overridable build-time macro?

dgarske commented on wolfcrypt/src/logging.c:
```diff
@@ -364,6 +364,27 @@ static void wolfssl_log(const int logLevel, const char *const logMessage)
+/* tell clang argument 1 is format */
+__attribute__((__format__ (__printf__, 1, 0)))
+#endif
+#if !defined(_WIN32) && defined(XVSNPRINTF)
```
It looks like types.h always defines `XVSNPRINTF`. Consider having another way to disable? Also if disabled should the format arg still be sent to `wolfssl_log`?

dgarske commented on src/ssl.c:
```diff
@@ -17989,14 +17995,17 @@ size_t wolfSSL_get_client_random(const WOLFSSL* ssl, unsigned char* out,
         InitX509(&ssl->peerCert, 0, ssl->heap);
 #endif
 
+        wolfSSL_SESSION_free(ssl->session);
```
Good fix!

julek-wolfssl replied:
Actually this should only be done if the session is invalid ie: the handshake didn't complete.

julek-wolfssl committed:
Only clear session when we didn't complete a handshake

- Allow overriding buffer size with `WOLFSSL_MSG_EX_BUF_SZ`
- Allow disabling `WOLFSSL_MSG_EX` by defining `NO_WOLFSSL_MSG_EX`
